core/arch/arm/tee/entry_std.c

```diff
@@ -152,122 +160,174 @@ static void copy_out_param(struct tee_ta_param *ta_param, uint32_t num_params,
-		return false;
+	if (num_params < 2)
+		return TEE_ERROR_BAD_PARAMETERS;
 
-	if (params[*num_meta].attr != req_attr ||
-	    params[*num_meta + 1].attr != req_attr)
-		return false;
+	if (params[0].attr != req_attr || params[1].attr != req_attr)
+		return TEE_ERROR_BAD_PARAMETERS;
 
-	tee_uuid_from_octets(uuid, (void *)&params[*num_meta].u.value);
-	tee_uuid_from_octets(&clnt_id->uuid,
-			     (void *)&params[*num_meta + 1].u.value);
-	clnt_id->login = params[*num_meta + 1].u.value.c;
+	tee_uuid_from_octets(uuid, (void *)&params[0].u.value);
+	clnt_id->login = params[1].u.value.c;
+	switch (clnt_id->login) {
+	case TEE_LOGIN_PUBLIC:
+		memset(&clnt_id->uuid, 0, sizeof(clnt_id->uuid));
+		break;
+	case TEE_LOGIN_USER:
+	case TEE_LOGIN_GROUP:
+	case TEE_LOGIN_APPLICATION:
+	case TEE_LOGIN_APPLICATION_USER:
+	case TEE_LOGIN_APPLICATION_GROUP:
+		tee_uuid_from_octets(&clnt_id->uuid,
+				     (void *)&params[1].u.value);
+		break;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
 
-	(*num_meta) += 2;
-	return true;
+	*num_meta = 2;
+	return TEE_SUCCESS;
 }
 
-static void entry_open_session(struct thread_smc_args *smc_args,
-			struct optee_msg_arg *arg, uint32_t num_params)
+static void inject_entropy_with_timestamp(void)
 {
-	struct tee_dispatch_open_session_in in;
-	struct tee_dispatch_open_session_out out;
-	struct optee_msg_param *params = OPTEE_MSG_GET_PARAMS(arg);
-	size_t num_meta = 0;
-
-	if (!get_open_session_meta(arg, num_params, &num_meta, &in.uuid,
-				   &in.clnt_id))
-		goto bad_params;
-
-	if (!copy_in_params(params + num_meta, num_params - num_meta,
-			    &in.param))
-		goto bad_params;
-
-	(void)tee_dispatch_open_session(&in, &out);
+	TEE_Time current;
 
-	copy_out_param(&out.param, num_params - num_meta, params + num_meta);
-
-	arg->session = (vaddr_t)out.sess;
-	arg->ret = out.msg.res;
-	arg->ret_origin = out.msg.err;
-	smc_args->a0 = OPTEE_SMC_RETURN_OK;
-	return;
+	if (tee_time_get_sys_time(&current) == TEE_SUCCESS)
+		tee_prng_add_entropy((uint8_t *)&current, sizeof(current));
+}
 
-bad_params:
-	DMSG("Bad params");
-	arg->ret = TEE_ERROR_BAD_PARAMETERS;
-	arg->ret_origin = TEE_ORIGIN_TEE;
+static void entry_open_session(struct thread_smc_args *smc_args,
+			       struct optee_msg_arg *arg, uint32_t num_params)
+{
+	TEE_Result res;
+	struct optee_msg_param *params = OPTEE_MSG_GET_PARAMS(arg);
+	TEE_ErrorOrigin err_orig = TEE_ORIGIN_TEE;
+	struct tee_ta_session *s = NULL;
+	TEE_Identity clnt_id;
+	TEE_UUID uuid;
+	struct tee_ta_param param;
+	size_t num_meta;
+
+	res = get_open_session_meta(num_params, params, &num_meta, &uuid,
+				    &clnt_id);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	res = copy_in_params(params + num_meta, num_params - num_meta, &param);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	res = tee_ta_open_session(&err_orig, &s, &tee_open_sessions, &uuid,
+				  &clnt_id, TEE_TIMEOUT_INFINITE, &param);
+	if (res != TEE_SUCCESS)
+		s = NULL;
+	copy_out_param(&param, num_params - num_meta, params + num_meta);
```

Comment:
This changes from original code: copy output param if open failed (and inject entropy).
Yet change seems ok.